Cabal/src/Distribution/Simple/SetupHooks/Rule.hs

```diff
@@ -322,13 +344,60 @@ dynamicRule dict depsCmd action dep res =
 -- consisting of a base directory and of a file path relative to that base
 -- directory path.
 --
--- In practice, this will be something like @( dir, toFilePath modName )@,
+-- In practice, this will be something like @'Location' dir ('moduleNameSymbolicPath' mod <.> "hs")@,
 -- where:
 --
 --  - for a file dependency, @dir@ is one of the Cabal search directories,
 --  - for an output, @dir@ is a directory such as @autogenComponentModulesDir@
 --    or @componentBuildDir@.
-type Location = (FilePath, FilePath)
+data Location where
+  Location
+    :: { locationBaseDir :: !(SymbolicPath Pkg (Dir baseDir))
+        -- ^ Base directory.
+       , locationRelPath :: !(RelativePath baseDir File)
+        -- ^ File path relative to base directory (including file extension).
+       }
+    -> Location
+
+instance Eq Location where
+  Location b1 l1 == Location b2 l2 =
+    (getSymbolicPath b1 == getSymbolicPath b2)
+      && (getSymbolicPath l1 == getSymbolicPath l2)
+instance Ord Location where
+  compare (Location b1 l1) (Location b2 l2) =
+    compare
+      (getSymbolicPath b1, getSymbolicPath l1)
+      (getSymbolicPath b2, getSymbolicPath l2)
+instance Binary Location where
+  put (Location base loc) = put (base, loc)
+  get = Location <$> get <*> get
+instance Structured Location where
+  structure _ =
+    Structure
+      tr
+      0
+      (show tr)
+      [
+        ( "Location"
+        ,
+          [ nominalStructure $ Proxy @(SymbolicPath Pkg (Dir (Tok "baseDir")))
+          , nominalStructure $ Proxy @(RelativePath (Tok "baseDir") File)
```

Comment:
Just a sanity check: is that right that `"baseDir"` is used in both cases? I'd expect `"relPath"` or something like that in the second case.

Reply:
Yes, it's a pair of file paths which "meet in the middle", i.e. a path from the package directory to some abstract base directory, and then a second path from that base directory to a file. That's the whole point of the GADT structure of this datatype.

Reply:
@sheaf that's how I understand this datatype too. But the string potion of this list, "baseDir", describes only the first portion of the path. That's why I thought there should be another string (like "relPath") to describe the second. In the current version, both potions are described as "baseDir", which is confusing to me. But I'm not well-versed in how Structured works so maybe I'm missing the point. Anyways, you have enough approvals to move forward now.

Reply:
The types are of the form `SymbolicPath from to` and `RelativePath from to`.

In this case we have:
  - base part: `from = Pkg`, `to = Dir baseDir`
  - relative part: `from = baseDir`, `to = File`.
  
`Dir Pkg --> Dir baseDir --> File`

The `SymbolicPath` abstraction does not keep track of what kind of file a path is, the type argument for any file is simply `File`. This design was motivated by the fact that tracking which kind of file something points to was just extra busywork that didn't buy any safety in practice.

Reply:
You seem to be explaining how `Location` is defined but I'm not asking this because we seem to agree on its meaning. My question is about the `Structured` instance. Let me quote it again here:
```haskell
  structure _ =
    Structure
      tr
      0
      (show tr)
      [
        ( "Location"
        ,
          [ nominalStructure $ Proxy @(SymbolicPath Pkg (Dir (Tok "baseDir")))
          , nominalStructure $ Proxy @(RelativePath (Tok "baseDir") File)      -- ***
sheaf marked this conversation as resolved.
          ]
        )
      ]
```
Do you see the line marked with `***`? What I don't understand about it is: while the object being constructed uses the `RelativePath` type, the `Tok` "marker" (I think of it as a marker, and correct me if I'm wrong) says "baseDir". Base directory is not expressed with `RealtivePath` at all, is it?

Reply:
Oh, scratch my previous comment. I think I got it now: this `Tok` business describes the type parameters of `SymbolicPath` / `RelativePath`...

Reply:
Yes, `Tok` is a bit of a hack to enable writing `Structure` instances for GADTs. I use `Tok` to generate names referring to existential variables.